flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/strategy/LazyFromSourcesSchedulingStrategy.java

```diff
@@ -0,0 +1,296 @@
+			}
+
+			if (shouldSchedule) {
+				executionVertexDeploymentOptions.add(
+					new ExecutionVertexDeploymentOption(executionVertexId, deploymentOptions.get(executionVertexId)));
+			}
+		}
+		schedulerOperations.allocateSlotsAndDeploy(executionVertexDeploymentOptions);
+	}
+
+	@Override
+	public void onExecutionStateChange(ExecutionVertexID executionVertexId, ExecutionState executionState) {
+		final SchedulingExecutionVertex schedulingVertex = schedulingTopology.getVertex(executionVertexId)
+			.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+
+		if (ExecutionState.FINISHED.equals(executionState)) {
+			List<SchedulingResultPartition> finishedPartitions = null;
+			for (SchedulingResultPartition srp : schedulingVertex.getProducedResultPartitions()) {
+				if (srp.getPartitionType().isBlocking()) {
+					SchedulingIntermediateDataSet intermediateDataSet = intermediateDataSets.get(srp.getResultId());
+					if (intermediateDataSet == null) {
+						throw new IllegalStateException("can not find scheduling intermediate dataset for "
+							+ srp.getResultId());
+					}
+					if (intermediateDataSet.decrementNumberOfRunningProducers()) {
+						if (finishedPartitions == null) {
+							finishedPartitions = new ArrayList<>();
+						}
+						finishedPartitions.addAll(intermediateDataSet.getSchedulingResultPartition());
+					}
+				}
+			}
+
+			if (finishedPartitions != null) {
+				List<SchedulingExecutionVertex> toScheduleConsumers = new ArrayList<>();
+				for (SchedulingResultPartition finishedPartition : finishedPartitions) {
+					toScheduleConsumers.addAll(finishedPartition.getConsumers());
+				}
+				scheduleBlockingConsumers(toScheduleConsumers);
+			}
+		}
+	}
+
+	@Override
+	public void onPartitionConsumable(ExecutionVertexID executionVertexId, ResultPartitionID resultPartitionId) {
+		final SchedulingExecutionVertex producerVertex = schedulingTopology.getVertex(executionVertexId)
+			.orElseThrow(() -> new IllegalStateException("can not find scheduling vertex for " + executionVertexId));
+		final SchedulingResultPartition resultPartition = schedulingTopology
+			.getResultPartition(resultPartitionId.getPartitionId())
+			.orElseThrow(() -> new IllegalStateException("can not find scheduling result partition for "
+				+ resultPartitionId));
+
+		if (!producerVertex.getProducedResultPartitions().contains(resultPartition)) {
+			throw new IllegalStateException("partition " + resultPartitionId
+				+ " is not the produced partition of " + executionVertexId);
+		}
+
+		List<ExecutionVertexDeploymentOption> executionVertexDeploymentOptions = new ArrayList<>();
+		for (SchedulingExecutionVertex schedulingVertex : resultPartition.getConsumers()) {
+			if (ExecutionState.CREATED.equals(schedulingVertex.getState()) &&
+				resultPartition.getPartitionType().isPipelined() &&
+				(PRODUCING.equals(resultPartition.getState()) || DONE.equals(resultPartition.getState()))) {
```

Comment:
I would say the contract of `onPartitionConsumable` should be that if the method is called, the partition has data, i.e., it’s `PRODUCING` or `DONE`.

Reply:
Great, the code could be more simple if having the contract .